src/app/shared/table-of-contents/table-of-contents.ts

```diff
@@ -0,0 +1,177 @@
+import {
+  Component, Input, AfterViewInit, Inject, ElementRef, OnInit
+} from '@angular/core';
+import {DOCUMENT} from '@angular/platform-browser';
+import {Router, ActivatedRoute, Event, NavigationEnd} from '@angular/router';
+import {Observable} from 'rxjs/Observable';
+import {Subscription} from 'rxjs/Subscription';
+import 'rxjs/add/observable/fromEvent';
+import 'rxjs/add/operator/debounceTime';
+
+interface Link {
+  /* id of the section*/
+  id: string;
+
+  /* header type h3/h4 */
+  type: string;
+
+  /* If the anchor is in view of the page */
+  active: boolean;
+
+  /* name of the anchor */
+  name: string;
+
+  /* top offset px of the anchor */
+  top: number;
+}
+
+@Component({
+  selector: 'table-of-contents',
+  styleUrls: ['./table-of-contents.scss'],
+  template: `
+    <div *ngIf="links?.length" class="docs-toc-container">
+      <div class="docs-toc-heading">Contents</div>
+      <nav>
+        <a [href]="_rootUrl + '#' + link.id"
+          *ngFor="let link of links; let i = index"
+          class="docs-level-{{link.type}}"
+          [class.active]="link.active">
+          {{link.name}}
+        </a>
+      </nav>
+    </div>
+  `
+})
+export class TableOfContents implements OnInit, AfterViewInit {
+
+  @Input() links: Link[] = [];
+  @Input() container: string;
+  @Input() headerSelectors = '.docs-markdown-h3,.docs-markdown-h4';
+
+  _rootUrl: string;
+  private _scrollContainer: any;
+  private _scrollSubscription: Subscription;
+  private _routeSubscription: Subscription;
+  private _fragmentObserver: MutationObserver;
+  private _headerObserver: MutationObserver;
+
+  private get scrollOffset(): number {
+    const {top} = this._element.nativeElement.getBoundingClientRect();
+    if (typeof this._scrollContainer.scrollTop !== 'undefined') {
+      return this._scrollContainer.scrollTop + top;
+    } else if (typeof this._scrollContainer.pageYOffset !== 'undefined') {
+      return this._scrollContainer.pageYOffset + top;
+    }
+  }
+
+  constructor(private _router: Router,
+              private _route: ActivatedRoute,
+              private _element: ElementRef,
+              @Inject(DOCUMENT) private _document: any) {
+
+    this._routeSubscription = this._router.events.subscribe((event: Event) => {
+      if (event instanceof NavigationEnd) {
+        const rootUrl = _router.url.split('#')[0];
+        if (rootUrl !== this._rootUrl) {
+          this.links.splice(0, this.links.length);
+          this.createLinks();
+          this._rootUrl = rootUrl;
+        }
+      }
+    });
+  }
+
+  ngOnInit(): void {
+    this._scrollContainer = this.container ?
+      this._document.querySelectorAll(this.container)[0] : window;
```

Comment:
the body element doesn't have a height set ![img](https://content.screencast.com/users/amcdaniel22/folders/Snagit/media/2f2e8d93-9a9f-4091-9ff8-9116abe3cb0a/2017-08-07_17-27-14.png)